torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -90,6 +90,66 @@ namespace nn {
 /// \endrst
 class SequentialImpl : public Cloneable<SequentialImpl> {
  public:
+  class NamedSubmodule {
+   public:
+    /// We don't allow constructing an empty `NamedSubmodule`.
+    NamedSubmodule() = delete;
+
+    /// Creates a `NamedSubmodule` from a (boxed) `Module`.
+    template <typename ModuleType>
+    NamedSubmodule(std::string name, std::shared_ptr<ModuleType> module_ptr) {
+      // Nesting Sequential doesn't work because `forward()`'s return type is
+      // templatized, so it'll give a nasty compiler error.
+      static_assert(
+          !std::is_same<SequentialImpl, ModuleType>::value,
+          "Sequential is not nestable");
+      static_assert(
+          torch::detail::is_module<ModuleType>::value,
+          "Can only add objects derived from nn::Module to Sequential");
+      static_assert(
+          torch::detail::has_forward<ModuleType>::value,
+          "Can only add modules with a forward() method to Sequential");
+      new (this) NamedSubmodule(std::move(name), AnyModule(std::move(module_ptr)));
```

Comment:
you don't want to do this -- it's extremely fragile.  Think about what happens if you allocate memory earlier or if we ever subclass these.

just use NamedSubmodule(...):NamedSubmodule(), which should be fine, because all the code here is static_asserts.

You have this a few places.